tests/integrations/subprocess/test_Subprocess__stream_output.py

```diff
@@ -30,42 +38,44 @@ def popen_process():
 def test_output(mock_sub, popen_process, capsys):
     """Readline output is printed."""
     mock_sub.stream_output("testing", popen_process)
+
     assert capsys.readouterr().out == ("output line 1\n" "\n" "output line 3\n")
+    mock_sub.cleanup.assert_called_once_with("testing", popen_process)
 
 
 def test_output_debug(mock_sub, popen_process, capsys):
     """Readline output is printed; debug mode should not add extra output."""
     mock_sub.command.logger = Log(verbosity=2)
 
     mock_sub.stream_output("testing", popen_process)
+
     assert capsys.readouterr().out == ("output line 1\n" "\n" "output line 3\n")
+    mock_sub.cleanup.assert_called_once_with("testing", popen_process)
 
 
 def test_output_deep_debug(mock_sub, popen_process, capsys):
     """Readline output is printed with debug return code in deep debug mode."""
     mock_sub.command.logger = Log(verbosity=3)
 
     mock_sub.stream_output("testing", popen_process)
+
     assert capsys.readouterr().out == (
         "output line 1\n" "\n" "output line 3\n" ">>> Return code: -3\n"
     )
+    mock_sub.cleanup.assert_called_once_with("testing", popen_process)
 
 
 def test_keyboard_interrupt(mock_sub, popen_process, capsys):
-    """Process is terminated if user sends CTRL+C."""
-    popen_process.stdout.readline.side_effect = [
-        "output line 1\n",
-        "\n",
-        KeyboardInterrupt(),
-    ]
-    mock_sub.cleanup = mock.MagicMock()
+    """KeyboardInterrupt is suppressed if user sends CTRL+C and all output is
+    printed."""
 
-    mock_sub.stream_output("testing", popen_process)
+    def send_ctrl_c():
+        raise KeyboardInterrupt()
 
-    # Response to the CTRL-C is a process cleanup.
-    mock_sub.cleanup.assert_called_once_with("testing", popen_process)
+    mock_sub.stream_output("testing", popen_process, stop_func=send_ctrl_c)
 
-    assert capsys.readouterr().out == ("output line 1\n" "\n")
+    assert capsys.readouterr().out == ("output line 1\n" "\n" "output line 3\n")
+    mock_sub.cleanup.assert_called_once_with("testing", popen_process)
```

Comment:
Moving the output streaming to a thread means that CTRL+C will no longer actually interrupt streaming since Python only handles signals in the main thread. Therefore, after a CTRL+C now, the code will request the Popen process terminates, wait 3 seconds, and then kill the process. This means everything that can be printed in 3 seconds will and control will then return to the user.

Reply:
The reason for the change makes sense, and I can't think of an alternative approach here (at least, not one that I actually like. You could set up a thread for testing purposes that sleeps for a time then sends a SIGINT to the main thread... but... ewwww... 🤮 )